### ISSUE TITLE
Numpy array operations differ in hypothesis vs manual tests

### ISSUE THREAD

**Issue by jordan-schneider** [HypothesisWorks/hypothesis#2480]
Hypothesis is reporting a test failure, but when I run the code manually (or when pytest runs it with fixed inputs), the error doesn't occur.

```python
import numpy as np
from hypothesis import given, note
from hypothesis.extra.numpy import array_shapes, arrays
from hypothesis.strategies import floats
from numpy.testing import assert_allclose

def normalize(log_weights: np.ndarray,) -> np.ndarray:
    weights = np.exp(log_weights)
    totals = np.sum(weights, axis=1)
    return (weights.T / totals).T

@given(
    log_weights=arrays(
        dtype=float,
        shape=array_shapes(min_dims=2, max_dims=2),
        elements=floats(0.000001, allow_nan=False, allow_infinity=False).map(np.log),
    )
)
def test_normalize(log_weights: np.ndarray):
    weights = normalize(log_weights)
    assert weights.shape == log_weights.shape
    note(f"weights={weights}")

    assert_allclose(np.sum(weights, axis=1), np.ones(weights.shape[0]))

def test_reported():
    weights = normalize(
        np.array([[-13.81551056, -13.81551056], [709.08956571, 709.08956571]])
    )
    assert_allclose(weights, np.array([[0.5, 0.5], [0.5, 0.5]]))
```

produces the following output

```
test/test_hypothesis_bug.py F.                                                                                                              [100%]

==================================================================== FAILURES =====================================================================
_________________________________________________________________ test_normalize __________________________________________________________________

    @given(
>       log_weights=arrays(
            dtype=float,
            shape=array_shapes(min_dims=2, max_dims=2),
            elements=floats(0.000001, allow_nan=False, allow_infinity=False).map(np.log),
        )
    )
    def test_normalize(log_weights: np.ndarray):

test/test_hypothesis_bug.py:16:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

log_weights = array([[-13.81551056, -13.81551056],
       [709.08956571, 709.08956571]])

    @given(
        log_weights=arrays(
            dtype=float,
            shape=array_shapes(min_dims=2, max_dims=2),
            elements=floats(0.000001, allow_nan=False, allow_infinity=False).map(np.log),
        )
    )
    def test_normalize(log_weights: np.ndarray):
        weights = normalize(log_weights)
        assert weights.shape == log_weights.shape
        note(f"weights={weights}")

>       assert_allclose(np.sum(weights, axis=1), np.ones(weights.shape[0]))
E       AssertionError:
E       Not equal to tolerance rtol=1e-07, atol=0
E
E       Mismatched elements: 1 / 2 (50%)
E       Max absolute difference: 1.
E       Max relative difference: 1.
E        x: array([1., 0.])
E        y: array([1., 1.])

test/test_hypothesis_bug.py:27: AssertionError
------------------------------------------------------------------- Hypothesis --------------------------------------------------------------------
Falsifying example: test_normalize(
    log_weights=array([[-13.81551056, -13.81551056],
           [709.08956571, 709.08956571]]),
)
weights=[[0.5 0.5]
 [0.  0. ]]
```

Note that the manual `test_reported` passes, and that the expected array is different from the one noted in test_normalize.

python: 3.7.7 [GCC 7.3.0] :: Anaconda, Inc. on linux
numpy: 1.20.0.dev0+9298eeb (current master as of writing)

**Comment by jordan-schneider** [comment]
It looks like the fundamental issue is that, inside the hypothesis test, some of the row sums are +inf, and then when we divide by inf we get 0. However, these same row sums in the manual test do not. I have checked that the numpy versions are the same in the two different tests, so either the small amount of rounding in the reported array is keeping us just below inf OR arrays generated by hypothesis are different than np.array outputs for some reason. I've already checked the dtypes, and all the relevant arrays are np.float64s.
